Warn if the getter is used, but the setter argument is not used

KingOfBrian opened the pull request:
This PR generates a warning if a setters getter is used, but the setter argument is not used. This is a common bug that occurs when a custom setter is implemented, but the getter is accidentally used instead of the setter argument.

This PR also removes some code that added method parameters to the checker. It was only used to warn about var parameters that were not mutated. Since var parameter support was dropped in 3.0, this was just busy work.

This was a rebase off of this [PR](https://github.com/apple/swift/pull/9258) that wasn't cooperating with CI.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-964](https://bugs.swift.org/browse/SR-964).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Commit:
Generate a warning if the implicit setter argument 'newValue' is not used.

Commit:
Change warning to only be reported if the parameter is unused but the getter is used.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke test

milseman commented:
@shahmishal is @swift-ci doing ok? It seems to bear a grudge against @KingOfBrian

shahmishal commented:
@swift-ci please smoke test

shahmishal commented:
@swift-ci please test

shahmishal commented:
@milseman It should be fixed now. 

@swift-ci please test

KingOfBrian commented:
Looks like one of the checks isn't reporting, and I don't think the two failures are related to this change. Really excited for this warning and would love to get it it 4.1 !

milseman commented:
CC @gottesmm @shahmishal that failure is the futimens one. In case you're still trying to track it down: https://ci.swift.org/job/swift-PR-osx/86/console

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/llvm/lib/Support/Unix/Path.inc:597:7: error: 'futimens' is only available on macOS 10.13 or newer [-Werror,-Wunguarded-availability-new]
```

milseman commented:
@swift-ci please test OS X Platform

marcelofabri commented:
Is something blocking this from being merged? I'd love to see it on 4.1 as well 🎉

milseman commented:
Did this ever get reviewed?

milseman commented:
CC @jrose-apple. If you have time, could you review this change?

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3701,6 +3701,10 @@ WARNING(pbd_never_used_stmtcond, none,
         (Identifier))
+WARNING(unused_setter_newvalue, none,
+        "setter argument %0 was never used, but the getter was accessed; "
+        "consider using %0 or removing %1",
```
I think the "removing" phrasing is confusing. How about something like "setter argument %0 was never used; did you mean to use it instead of accessing the property's current value"?

Ideally, we'd put the warning on one of the uses of the getter, and maybe even have a fix-it if it's just one use, but that's probably more work than you want to do right now.

jrose-apple reviewed (commented):
The implementation feels a bit funny to me, since that map is normally for *stored* properties, but I can see how it works nicely with the existing tracking. So my only remaining concern is with the phrasing of the diagnostic.

Commit:
Update error message

Commit:
Generate a fix-it to change the first getter usage into the setter parameter.

CodaFi commented:
Well, that’s not... quite a rebase.

KingOfBrian commented:
Hah! was hoping to fix this before anyone noticed. Me and git didn't get along last night. I think my initial rebase was off... and things escalated. I'll clean it up shortly!

KingOfBrian commented:
Thanks for the feedback @jrose-apple -- I added the fixit, and moved where the error was reported. I track the first `DeclRefExpr` to use the getter for the diagnostic, which let me drop the `RK_DoNotReport`.

jrose-apple commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -2093,7 +2093,11 @@ class VarDeclUsageChecker : public ASTWalker {
+
+  /// This is a mapping from a VarDecl property getter to the expression that
+  /// used it.
+  llvm::SmallDenseMap<VarDecl*, DeclRefExpr*> GetterDeclUsageMap;
```
This map will only ever have one entry, right?

KingOfBrian replied:
Yea -- sort of funny, I woke up this morning and that was literally my first thought. I converted it to a pair, but I'm not sure if that's a normal C++ construct to use. Would that make sense to you?

jrose-apple replied:
I think I would just use two instance variables, and not worry about keeping them tied together. You also set each half at a different point in execution anyway.

(Also, may as well make them `const`.)

jrose-apple commented on test/decl/var/usage.swift:
```diff
@@ -274,3 +274,34 @@ func sr2421() {
+  var suspiciousSetter: String {
+    get { return "" }
+    set {
+      print(suspiciousSetter) // expected-warning {{setter argument 'newValue' was never used; did you mean to use it instead of accessing the property's current value?}} expected-note {{do you want to replace 'suspiciousSetter' with 'newValue'?}}
```
If you're going to have the note, I think the "did you mean" part belongs in the note, rather than describing the fix-it. Also, you can test fix-its with the expected-* syntax! Use `{{1-2=foo}}` syntax after the diagnostic text to mean "replace columns 1 to 2 with 'foo'".

KingOfBrian replied:
Yea, that makes sense. I thought a note was more appropriate since the fixit isn't always the correct course of action. I can just do the warning if you think that makes more sense.

Otherwise, does this language look good?
```
warning: setter argument 'newValue' was never used, but the property was accessed
note: did you mean to use 'newValue' instead of accessing the property's current value?
```

jrose-apple replied:
Sure, that seems fine, and I agree that the separate note is a good idea.

Commit:
Clean up implementation

Commit:
Improve diagnostic language

jrose-apple commented:
@swift-ci Please test
